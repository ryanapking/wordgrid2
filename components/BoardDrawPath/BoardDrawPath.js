import React from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

const BoardDrawPath = props => {
  const { square1, square2, boardLocation } = props;

  const height = boardLocation.rowHeight * .1;
  const borderRadius = height / 2;

  const p1 = getRelativeCoordinates(boardLocation, square1.rowIndex, square1.columnIndex);
  const p2 = getRelativeCoordinates(boardLocation, square2.rowIndex, square2.columnIndex);
  const angle = calculateAngle(p1, p2);
  const length = calculateLength(p1, p2) + borderRadius;
  const midpoint = calculateMidPoint(p1, p2);

  const style = {
    left: midpoint.x - (length / 2),
    top: midpoint.y - (height / 2),
    width: length,
    transform: [{rotate: angle + 'deg'}],
    height: height,
    borderRadius: borderRadius,
    backgroundColor: '#5283ff',
    position: 'absolute'
  };

  return (
    <View style={style} />
  );
};

BoardDrawPath.propTypes = {
  square1: PropTypes.shape({
    rowIndex: PropTypes.number,
    columnIndex: PropTypes.number,
  }),
  square2: PropTypes.shape({
    rowIndex: PropTypes.number,
    columnIndex: PropTypes.number,
  }),
  boardLocation: PropTypes.shape({
    rowHeight: PropTypes.number,
    columnWidth: PropTypes.number,
  }),
};

const getRelativeCoordinates = (boardLocation, rowIndex, columnIndex) => {
  const y = (rowIndex + .5) * boardLocation.rowHeight;
  const x = (columnIndex + .5) * boardLocation.columnWidth;
  return { x, y };
};

const calculateLength = (p1, p2) => {
  return Math.hypot(p1.x - p2.x, p1.y - p2.y);
};

const calculateMidPoint = (p1, p2) => {
  const x = (p1.x + p2.x) / 2;
  const y = (p1.y + p2.y) / 2;
  return {x, y};
};

const calculateAngle = (p1, p2) => {
  let dy = p1.y - p2.y;
  let dx = p1.x - p2.x;
  let theta = Math.atan2(dy, dx); // range (-PI, PI]
  theta *= 180 / Math.PI; // rads to degs, range (-180, 180]
  //if (theta < 0) theta = 360 + theta; // range [0, 360)
  return theta;
};

export default BoardDrawPath;