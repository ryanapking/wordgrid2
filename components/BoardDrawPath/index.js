/* istanbul ignore file */
export { default } from "./BoardDrawPath";