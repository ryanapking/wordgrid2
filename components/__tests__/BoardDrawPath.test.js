import React from 'react';
import { render } from 'react-native-testing-library';

import BoardDrawPath from "../BoardDrawPath";

const testProps = {
  square1: {rowIndex: 7, columnIndex: 5, letter: "s"},
  square2: {rowIndex: 6, columnIndex: 4, letter: "m"},
  boardLocation: {rowHeight:38.6, columnWidth: 37.5},
};

describe('BoardDrawPath', () => {
  it('renders correctly', () => {
    const { unmount, toJSON } = render(<BoardDrawPath square1={testProps.square1} square2={testProps.square2} boardLocation={testProps.boardLocation} />);
    expect(toJSON()).toMatchSnapshot();
    unmount();
  });
});