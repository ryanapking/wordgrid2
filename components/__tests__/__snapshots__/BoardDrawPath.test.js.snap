// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardDrawPath renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#5283ff",
      "borderRadius": 1.9300000000000002,
      "height": 3.8600000000000003,
      "left": 159.62677634997064,
      "position": "absolute",
      "top": 268.27,
      "transform": Array [
        Object {
          "rotate": "45.828133621083396deg",
        },
      ],
      "width": 55.74644730005874,
    }
  }
/>
`;
